usr.man/man3/fnmatch.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Guido van Rossum.
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)fnmatch.3	8.3 (Berkeley) 4/28/95
.\" $Id: fnmatch.3,v 1.2 1999/11/22 07:42:06 stever Exp $
.\"
.TH FNMATCH 3 "22 February 1996" GNO "Library Routines"
.SH NAME
.BR fnmatch
\- match filename or pathname
.SH SYNOPSIS
#include <fnmatch.h>
.sp 1
int
\fBfnmatch\fR (const char *\fIpattern\fR, const char *\fIstring\fR,
int \fIflags\fR);
.SH DESCRIPTION
The
.BR fnmatch 
function
matches patterns according to the rules used by the shell.
It checks the string specified by the
.I string
argument to see if it matches the pattern specified by the
.I pattern
argument.
.LP
The
.I flags
argument modifies the interpretation of
.I pattern
and
.IR string .
The value of
.I flags
is the bitwise inclusive
.BR OR
of any of the following
constants, which are defined in the include file <fnmatch.h>.
.RS
.IP FNM_NOESCAPE
Normally, every occurrence of a backslash
.BR \e
followed by a character in
.I pattern
is replaced by that character.
This is done to negate any special meaning for the character.
If the
.BR FNM_NOESCAPE
flag is set, a backslash character is treated as an ordinary character.
.IP FNM_PATHNAME
Slash characters in
.I string
must be explicitly matched by slashes in
.IR pattern .
If this flag is not set, then slashes are treated as regular characters.
.IP FNM_PERIOD
Leading periods in
.I string
must be explicitly matched by periods in
.IR pattern .
If this flag is not set, then leading periods are treated as regular
characters.
The definition of ``leading'' is related to the specification of
.BR FNM_PATHNAME.
A period is always ``leading'' if it is the first character in
.BR string .
Additionally, if
.BR FNM_PATHNAME
is set,
a period is ``leading'' if it immediately follows a slash.
.IP FNM_CASEFOLD
If this flag is set, then the filename matching will be case insensitive.
This is an extension for GNO.
.IP FNM_LEADING_DIR
Ignore any trailing sequence
.B /*
after successfully matching
.IR pattern .
This is a GNU extension.
.RE
.LP
In the GNO implementation, if the
.B FNM_PATHNAME
flag is set, any colon
.RB ( : )
characters appearing in
.IR pattern
or
.IR string
will be mapped to slash
.RB ( / )
characters before any match is attempted, unless either
.IR pattern
or 
.IR string
contain both colons and slashes.
.SH RETURN VALUES
The
.BR fnmatch 
function returns zero if
.I string
matches the pattern specified by
.IR pattern ,
otherwise, it returns the value
.BR FNM_NOMATCH .
.SH BUGS
The pattern
.BR *
matches the empty string, even if
.BR FNM_PATHNAME
is specified.
.LP
This implementation of 
.BR fnmatch
uses recursion.  While this is not strictly a bug, it is a serious
limitation on the IIgs where stack space is at a premium.
.SH STANDARDS
Other than the mapping of 
.BR :
characters to 
.BR /
characters, the
.BR fnmatch 
function conforms to POSIX 1003.2.
The
.BR FNM_CASEFOLD
and
.B FNM_LEADING_DIR
flags are extensions to the standard.
.SH HISTORY
The
.BR fnmatch 
function first appeared in 4.4BSD.
.SH SEE ALSO
.BR sh (1),
.BR glob (3),
.BR regex (3)


lib/libc/gen/fnmatch.c
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Guido van Rossum.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * This implementation uses recursion.  It should be rewritten to avoid
 * it due to stack limitations on the IIgs.
 *
 * $Id: fnmatch.c,v 1.1 1997/02/28 05:12:44 gdr Exp $
 *
 * This file is formatted for tab stops every 8 characters.
 */
 
#ifdef __ORCAC__
segment "libc_gen__";
#endif

#pragma optimize 0
#pragma debug 0
#pragma memorymodel 0

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)fnmatch.c	8.2 (Berkeley) 4/16/94";
#endif /* LIBC_SCCS and not lint */

/*
 * Function fnmatch() as specified in POSIX 1003.2-1992, section B.6.
 * Compares a filename or pathname to a pattern.
 */

#include <fnmatch.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <err.h>

#define	EOS	'\0'

static const char *rangematch __P((const char *, int, int));
static void _fnmatch_map (const char *, const char *, char **, char **, int);

int
fnmatch(const char *opattern, const char *ostring, int flags)
{
	const char *stringstart;
	char c, test;
	char *pattern, *string;

	_fnmatch_map(ostring, opattern, &string, &pattern, flags);

	for (stringstart = string;;)
		switch (c = *pattern++) {
		case EOS:
			return (*string == EOS ? 0 : FNM_NOMATCH);
		case '?':
			if (*string == EOS)
				return (FNM_NOMATCH);
			if (*string == '/' && (flags & FNM_PATHNAME))
				return (FNM_NOMATCH);
			if (*string == '.' && (flags & FNM_PERIOD) &&
			    (string == stringstart ||
			    ((flags & FNM_PATHNAME) && *(string - 1) == '/')))
				return (FNM_NOMATCH);
			++string;
			break;
		case '*':
			c = *pattern;
			/* Collapse multiple stars. */
			while (c == '*')
				c = *++pattern;

			if (*string == '.' && (flags & FNM_PERIOD) &&
			    (string == stringstart ||
			    ((flags & FNM_PATHNAME) && *(string - 1) == '/')))
				return (FNM_NOMATCH);

			/* Optimize for pattern with * at end or before /. */
			if (c == EOS)
				if (flags & FNM_PATHNAME)
					return (strchr(string, '/') == NULL ?
					    0 : FNM_NOMATCH);
				else
					return (0);
			else if (c == '/' && flags & FNM_PATHNAME) {
				if ((string = strchr(string, '/')) == NULL)
					return (FNM_NOMATCH);
				break;
			}

			/* General case, use recursion. */
			while ((test = *string) != EOS) {
				if (!fnmatch(pattern, string, flags & ~FNM_PERIOD))
					return (0);
				if (test == '/' && flags & FNM_PATHNAME)
					break;
				++string;
			}
			return (FNM_NOMATCH);
		case '[':
			if (*string == EOS)
				return (FNM_NOMATCH);
			if (*string == '/' && flags & FNM_PATHNAME)
				return (FNM_NOMATCH);
			if ((pattern =
			    rangematch(pattern, *string, flags)) == NULL)
				return (FNM_NOMATCH);
			++string;
			break;
		case '\\':
			if (!(flags & FNM_NOESCAPE)) {
				if ((c = *pattern++) == EOS) {
					c = '\\';
					--pattern;
				}
			}
			/* FALLTHROUGH */
		default:
			if (c != *string++)
				return (FNM_NOMATCH);
			break;
		}
	/* NOTREACHED */
}

static const char *
rangematch(const char *pattern, int test, int flags)
{
	int negate, ok;
	char c, c2;

	/*
	 * A bracket expression starting with an unquoted circumflex
	 * character produces unspecified results (IEEE 1003.2-1992,
	 * 3.13.2).  This implementation treats it like '!', for
	 * consistency with the regular expression syntax.
	 * J.T. Conklin (conklin@ngai.kaleida.com)
	 */
	if (negate = (*pattern == '!' || *pattern == '^'))
		++pattern;

	for (ok = 0; (c = *pattern++) != ']';) {
		if (c == '\\' && !(flags & FNM_NOESCAPE))
			c = *pattern++;
		if (c == EOS)
			return (NULL);
		if (*pattern == '-'
		    && (c2 = *(pattern+1)) != EOS && c2 != ']') {
			pattern += 2;
			if (c2 == '\\' && !(flags & FNM_NOESCAPE))
				c2 = *pattern++;
			if (c2 == EOS)
				return (NULL);
			if (c <= test && test <= c2)
				ok = 1;
		} else if (c == test)
			ok = 1;
	}
	return (ok == negate ? NULL : pattern);
}

static void
_fnmatch_map (const char *opath, const char *orex, char **npath, char **nrex,
	      int flags) {
	static char *path = NULL;
	static char *rex = NULL;
	char *p;
	int i;
	
	/*
	 * create copies of opath and orex; this depends on an
	 * ANSI implementation of realloc (accepts NULL pointer)
	 */
	path = realloc(path, strlen(opath) + 1);
	rex = realloc(rex, strlen(orex) + 1);
	if (path == NULL || rex == NULL) {
		err (1, "fnmatch could not allocate internal buffer");
		/*NOTREACHED*/
	}
	strcpy(path, opath);
	strcpy(rex, orex);

	/* fold case if necessary */
	if (flags & FNM_CASEFOLD) {
		for (p = path; *p != '\0'; p++) {
			if (isupper(*p)) {
				*p = _tolower(*p);
			}
		}
		for (p = rex; *p != '\0'; p++) {
			if (isupper(*p)) {
				*p = _tolower(*p);
			}
		}
	}

	/*
	 * If either pattern or string contain _both_ a ':' and a '/',
	 * then we leave them exactly as they are.  Otherwise, all colons
	 * are mapped to '/'.
	 */
	if (!((strchr(path, ':') && strchr(path, '/')) ||
	     (strchr(rex,  ':') && strchr(rex,  '/')))) {
		for (i=0; i<2; i++) {
			switch (i) {
			case 0:
				p = path;
				break;
			case 1:
				p = rex;
				break;
			}
			while (*p) {
				if (*p == ':') {
					*p = '/';
				}
				p++;
			}
		}
	}

	/* give the caller pointers to our buffer */	     
	*npath = path;
	*nrex = rex;
	return;
}
